connector/oauth/oauth.go

```diff
@@ -0,0 +1,286 @@
+package oauth
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"golang.org/x/oauth2"
+
+	"github.com/dexidp/dex/connector"
+	"github.com/dexidp/dex/pkg/log"
+)
+
+type oauthConnector struct {
+	clientID             string
+	clientSecret         string
+	redirectURI          string
+	tokenURL             string
+	authorizationURL     string
+	userInfoURL          string
+	scopes               []string
+	userIDKey            string
+	userNameKey          string
+	preferredUsernameKey string
+	emailKey             string
+	emailVerifiedKey     string
+	groupsKey            string
+	httpClient           *http.Client
+	logger               log.Logger
+}
+
+type connectorData struct {
+	AccessToken string
+}
+
+type Config struct {
+	ClientID           string   `json:"clientID"`
+	ClientSecret       string   `json:"clientSecret"`
+	RedirectURI        string   `json:"redirectURI"`
+	TokenURL           string   `json:"tokenURL"`
+	AuthorizationURL   string   `json:"authorizationURL"`
+	UserInfoURL        string   `json:"userInfoURL"`
+	Scopes             []string `json:"scopes"`
+	RootCAs            []string `json:"rootCAs"`
+	InsecureSkipVerify bool     `json:"insecureSkipVerify"`
+	UserIDKey          string   `json:"userIDKey"` // defaults to "id"
+	ClaimMapping       struct {
+		UserNameKey          string `json:"userNameKey"`          // defaults to "user_name"
+		PreferredUsernameKey string `json:"preferredUsernameKey"` // defaults to "preferred_username"
+		GroupsKey            string `json:"groupsKey"`            // defaults to "groups"
+		EmailKey             string `json:"emailKey"`             // defaults to "email"
+		EmailVerifiedKey     string `json:"emailVerifiedKey"`     // defaults to "email_verified"
+	} `json:"claimMapping"`
+}
+
+func (c *Config) Open(id string, logger log.Logger) (connector.Connector, error) {
+	var err error
+
+	if c.UserIDKey == "" {
+		c.UserIDKey = "id"
+	}
+
+	if c.ClaimMapping.UserNameKey == "" {
+		c.ClaimMapping.UserNameKey = "user_name"
+	}
+
+	if c.ClaimMapping.PreferredUsernameKey == "" {
+		c.ClaimMapping.PreferredUsernameKey = "preferred_username"
+	}
+
+	if c.ClaimMapping.GroupsKey == "" {
+		c.ClaimMapping.GroupsKey = "groups"
+	}
+
+	if c.ClaimMapping.EmailKey == "" {
+		c.ClaimMapping.EmailKey = "email"
+	}
+
+	if c.ClaimMapping.EmailVerifiedKey == "" {
+		c.ClaimMapping.EmailVerifiedKey = "email_verified"
+	}
```

Comment:
Changing the struct itself is not very elegant. Using local variables would be nicer, but this is a nit.